File: api/routers/fitness.py
"""Fitness API router: Google Health API v4 integration endpoints."""
import logging

from fastapi import APIRouter, Depends, HTTPException, Query
from slowapi import Limiter
from slowapi.util import get_remote_address

from api.config import get_config
from api.dependencies import get_current_user_id
from api.schemas.fitness import (
    FitnessConnectionStatus,
    FitnessDataResponse,
    FitnessSyncResponse,
    StoreTokensRequest,
)
from api.services.fitness_service import FitnessService

logger = logging.getLogger(__name__)
limiter = Limiter(key_func=get_remote_address)

DEFAULT_PROVIDER = "google_health"


def create_fitness_router(fitness_service: FitnessService) -> APIRouter:
    router = APIRouter(prefix="/fitness")

    @router.post("/tokens", status_code=204)
    def store_tokens(
        body: StoreTokensRequest,
        user_id: int = Depends(get_current_user_id),
    ):
        """Store OAuth tokens sent by the frontend after PKCE code exchange."""
        cfg = get_config()
        if not cfg.ENABLE_GOOGLE_HEALTH:
            raise HTTPException(status_code=404)
        try:
            fitness_service.store_tokens(
                user_id=user_id,
                provider=body.provider,
                access_token=body.access_token,
                refresh_token=body.refresh_token,
                expires_in=body.expires_in,
                cfg=cfg,
            )
        except Exception as exc:
            logger.error("Failed to store fitness tokens for user %s: %s", user_id, exc)
            raise HTTPException(status_code=500, detail="Failed to store tokens")

    @router.get("/status", response_model=FitnessConnectionStatus)
    def get_status(
        provider: str = Query(default=DEFAULT_PROVIDER),
        user_id: int = Depends(get_current_user_id),
    ):
        return fitness_service.get_connection_status(user_id, provider)

    @router.post("/sync", response_model=FitnessSyncResponse)
    def trigger_sync(
        provider: str = Query(default=DEFAULT_PROVIDER),
        days: int = Query(default=30, ge=1, le=90),
        user_id: int = Depends(get_current_user_id),
    ):
        cfg = get_config()
        if not cfg.ENABLE_GOOGLE_HEALTH:
            raise HTTPException(status_code=404)
        try:
            rows = fitness_service.sync(user_id, provider, cfg, days)
            return {"status": "ok", "rows_synced": rows, "provider": provider}
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc))
        except Exception as exc:
            logger.error("Fitness sync failed for user %s: %s", user_id, exc)
            raise HTTPException(status_code=500, detail="Sync failed")

    @router.delete("/disconnect", status_code=200)
    def disconnect(
        provider: str = Query(default=DEFAULT_PROVIDER),
        user_id: int = Depends(get_current_user_id),
    ):
        fitness_service.disconnect(user_id, provider)
        return {"status": "ok"}

    @router.get("/data", response_model=FitnessDataResponse)
    def get_data(
        provider: str = Query(default=DEFAULT_PROVIDER),
        start_date: str = Query(default=None),
        end_date: str = Query(default=None),
        user_id: int = Depends(get_current_user_id),
    ):
        data = fitness_service.get_data(user_id, start_date=start_date, end_date=end_date)
        return {"data": data, "provider": provider}

    return router

File: functions/_shared/routes/fitness.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { z } from 'zod';
import { HTTPException } from 'hono/http-exception';
import type { AppEnv } from '../lib/env';
import { isTruthy } from '../lib/env';
import { createDb } from '../db/client';
import { storeTokensRequestSchema } from '@shared/schemas/fitness';
import {
  FitnessError,
  disconnect,
  getConnectionStatus,
  getData,
  storeTokens,
  sync,
} from '../services/fitness-service';

const DEFAULT_PROVIDER = 'google_health';

const statusQuerySchema = z.object({ provider: z.string().default(DEFAULT_PROVIDER) });
const syncQuerySchema = z.object({
  provider: z.string().default(DEFAULT_PROVIDER),
  days: z.coerce.number().int().min(1).max(90).default(30),
});
const dataQuerySchema = z.object({
  provider: z.string().default(DEFAULT_PROVIDER),
  start_date: z.string().optional(),
  end_date: z.string().optional(),
});

export const fitnessRoutes = new Hono<AppEnv>();

fitnessRoutes.post('/fitness/tokens', zValidator('json', storeTokensRequestSchema), async (c) => {
  if (!isTruthy(c.env.ENABLE_GOOGLE_HEALTH)) {
    throw new HTTPException(404);
  }

  const userId = c.get('userId');
  const body = c.req.valid('json');
  const db = createDb(c.env.DB);

  try {
    await storeTokens(
      db,
      c.env,
      userId,
      body.provider,
      body.access_token,
      body.refresh_token,
      body.expires_in
    );
  } catch (err) {
    console.error(`Failed to store fitness tokens for user ${userId}:`, err);
    throw new HTTPException(500, { message: 'Failed to store tokens' });
  }

  return c.body(null, 204);
});

fitnessRoutes.get('/fitness/status', zValidator('query', statusQuerySchema), async (c) => {
  const userId = c.get('userId');
  const { provider } = c.req.valid('query');
  const db = createDb(c.env.DB);
  return c.json(await getConnectionStatus(db, userId, provider));
});

fitnessRoutes.post('/fitness/sync', zValidator('query', syncQuerySchema), async (c) => {
  if (!isTruthy(c.env.ENABLE_GOOGLE_HEALTH)) {
    throw new HTTPException(404);
  }

  const userId = c.get('userId');
  const { provider, days } = c.req.valid('query');
  const db = createDb(c.env.DB);

  try {
    const rows = await sync(db, c.env, userId, provider, days);
    return c.json({ status: 'ok', rows_synced: rows, provider });
  } catch (err) {
    if (err instanceof FitnessError) {
      throw new HTTPException(400, { message: err.message });
    }
    console.error(`Fitness sync failed for user ${userId}:`, err);
    throw new HTTPException(500, { message: 'Sync failed' });
  }
});

fitnessRoutes.delete('/fitness/disconnect', zValidator('query', statusQuerySchema), async (c) => {
  const userId = c.get('userId');
  const { provider } = c.req.valid('query');
  const db = createDb(c.env.DB);
  await disconnect(db, userId, provider);
  return c.json({ status: 'ok' });
});

fitnessRoutes.get('/fitness/data', zValidator('query', dataQuerySchema), async (c) => {
  const userId = c.get('userId');
  const { provider, start_date: startDate, end_date: endDate } = c.req.valid('query');
  const db = createDb(c.env.DB);
  const data = await getData(db, userId, startDate, endDate);
  return c.json({ data, provider });
});
